src/ForSyDeIRToProceduralIR.hs

```diff
@@ -0,0 +1,290 @@
+        else
+          if (elem id (systemOutputs initialContext))
+            then
+              let ioTokenStmt = SVarDecl TInt ("output_" ++ id)
+               in initialContext {ioTokens = ioTokenStmt : ioTokens initialContext, initBuffers = initStmt : initBuffers initialContext, freeBuffers = freeStmt : freeBuffers initialContext}
+            else
+              initialContext {initBuffers = initStmt : initBuffers initialContext, freeBuffers = freeStmt : freeBuffers initialContext}
+
+-- | Translates an `IRConstructor` into a set of statements which are used to
+-- update the `TranslationContext`.
+translateIRConstructor :: TranslationContext -> IRConstructor -> TranslationContext
+translateIRConstructor initialContext constructor = case constructor of
+  IRDelay _ tokens (inputSignal, outputSignal) ->
+    -- `IRDelay` results in a list of statements adding initial delay tokens
+    -- into a signal buffer.
+    let maybeBufferId = lookup outputSignal (delayBuffers initialContext)
+     in case maybeBufferId of
+          Just bufferId ->
+            let foldDelayTokens :: [Statement] -> Int -> [Statement]
+                foldDelayTokens acc token =
+                  let stmt = SExpr (ECall "write_token" [EVar bufferId, EInt token])
+                   in (stmt : acc)
+                delayStmts = (foldl' foldDelayTokens [] tokens)
+             in initialContext {initDelay = delayStmts ++ initDelay initialContext}
+          Nothing -> error ("translateIRConstructor - delay buffer not found for signals: " ++ inputSignal ++ ", " ++ outputSignal)
+  IRActor actorId actorType functionId (inputSignals, outputSignals) ->
+    -- `IRActor` results in a statement calling an SDF actor function within
+    -- source rates, target rates, and a function as arguments.
+    let sourceRates = map (getSourceRate initialContext) inputSignals
+        targetRates = map (getTargetRate initialContext) outputSignals
+        actorName = translateActorType actorType
+        translatedInputSignals = map (translateSignalId initialContext) inputSignals
+        translatedOutputSignals = map (translateSignalId initialContext) outputSignals
+        actorCallStmt =
+          SExpr
+            ( ECall
+                actorName
+                ( (map (\rate -> EInt (fromIntegral rate)) sourceRates)
+                    ++ (map (\rate -> EInt (fromIntegral rate)) targetRates)
+                    ++ (map (\id -> EVar id) translatedInputSignals)
+                    ++ (map (\id -> EVar id) translatedOutputSignals)
+                    ++ [EVar functionId]
+                )
+            )
+        inputStmts = foldl' foldActorSignals [] inputSignals
+        outputStmts = foldl' foldActorSignals [] outputSignals
+        stmts = reverse inputStmts ++ [actorCallStmt] ++ reverse outputStmts
```

Comment:
If you reverse after anyway it might make sense to use foldr instead which removes the need for that if you don't rely on the order of evaluation.

Reply:
I won't object to keeping it as `foldl'`, but `foldr` should not have the issue as the pure `foldl` where it requires stack space proportional to the length. `foldl'` might be a little more efficient than `foldr` since we're working with finite lists, but it shouldn't be a large one.